content/js1/sprints/4/prep/index.md

```diff
@@ -8,4 +8,666 @@ backlog= 'Module-JS1'
+  "scripts": {
+    "test": "jest"
+  },
+  "devDependencies": {
+    "jest": "^29.5.0"
+  }
+}
+```
+
+Finally, we'll need to run our tests.
+Now we can run the command `npm test`.
+
+When we execute the command, `npm test`, the command will look inside the "scripts" section of the package.json and look up the commmand for "test" - in this case, "jest".
+
+### Understanding test feedback
+
+We currently have a project structure like this:
+
+```raw
+week-4-test-example
+├── get-ordinal-number.test.js
+├── package.json
+├── package-lock.json
+└── node_modules
+
+1 directory, 3 files
+```
+
+And `get-ordinal-number.test.js`
+looks like this
+
+```js
+test("converts 1 to an ordinal number", function () {
+  expect(getOrdinalNumber(1)).toBe("1st");
+});
+```
+
+After running the test above, we should get feedback indicating whether or not the test has passed.
+
+{{<note type="exercise" title="exercise">}}
+Predict what the test feedback will say when the test above is executed.
+{{</note>}}
+
+### 🚢 Defining the function
+
+At the moment, our test feedback gives the following:
+
+![test-reference-error](test-reference-error.png)
+
+The test code is throwing a **ReferenceError**.
+
+We can define `getOrdinalNumber` in our test file.
+
+```js
+function getOrdinalNumber() {}
+
+test("converts 1 to an ordinal number", function () {
+  expect(getOrdinalNumber(1)).toBe("1st");
+});
+```
+
+Now we can run the tests again and check the test feedback.
+
+### Assertion errors
+
+We now get the following feedback:
+
+![test-feedback-fail](test-feedback-fail.png)
+
+Jest tells us 3 main things:
+
+1. The test case that failed
+2. The targetOutput and the currentOutput
+3. The line number where error occurred
+
+Jest defines **Expected** and **Received** in the test feedback:
+
+- Received: `undefined`
+- Expected: "1st"
+
+{{<note type="exercise" title="exercise">}}
+
+What are the values of Expected and Received in the test output?
+How do Received and Expected match up with the targetOutput and expectedOutput ?
+
+What line number did the test case fail on?
+
+{{</note>}}
+
+### Passing `getOrdinalNumber`
+
+We can now make the test pass by implementing functionality for the first test case.
+We could write the following:
+
+get-ordinal-number.test.js
+
+```js {linenos=table,hl_lines=["2"],linenostart=1}
+function getOrdinalNumber() {
+  return "1st";
+}
+
+test("converts 1 to an ordinal number", function () {
+  expect(getOrdinalNumber(1)).toBe("1st");
+});
+```
+
+### 🗄️ Generalising further
+
+In English, ordinal numbers mostly follow the same pattern.
+
+> Numbers ending in 1 will generally have an ordinal number ending in "st".
+
+Here are some examples of this pattern,
+
+1st, 11**th**, 21st, 31st, 41st,...
+
+All the numbers ending in 1 will continue to end in ` "st"``, with the exception of 11.
+11 is slightly different and ends with a  `"th"`.
+
+We can now create a test case to check that `getOrdinalNumber` works for _any_ number ending in 1.
```

Comment:
```suggestion
We can now update our test case to check that `getOrdinalNumber` works for lots of different numbers ending in 1.
```

If we're talking about "any", I'd think you're talking about property-based testing like using https://github.com/dubzzz/fast-check